src/ccl_cls.c

```diff
@@ -18,202 +18,271 @@ typedef struct{
+    double pk_ratio = fabs(ccl_f2d_t_eval(psp, lk, a_lp, cosmo, status) /
+                           ccl_f2d_t_eval(psp, lk, a_l, cosmo, status));
+
+    // Compute kernel and trasfer at chi_{l+1}
+    double w_p = ccl_cl_tracer_t_get_kernel(tr, chi_lp, status);
+    double t_p = ccl_cl_tracer_t_get_transfer(tr, lk,a_lp, status);
+
+    // sqrt(2l+1/2l+3)
+    double sqell = sqrt(lp1h*pk_ratio/lp3h);
+    if (tr->der_bessel == 1)
+      dd = l*w*t/lp1h-sqell*w_p*t_p;
     else //we assume der_bessel=2 here to avoid extra if clause
-      dd=sqell*2*w_p*t_p/lp3h-(0.25+2*l)*w*t/(lp1h*lp1h);
+      dd = sqell*2*w_p*t_p/lp3h - (0.25+2*l)*w*t/(lp1h*lp1h);
   }
   return dd*fl;
 }
 
-static double transfer_limber_wrap(double l,double lk,double k,double chi,double a,
-				   ccl_cl_tracer_collection_t *trc,
-				   ccl_cosmology *cosmo,ccl_f2d_t *psp,
-				   int *status)
-{
+static double transfer_limber_wrap(double l,double lk, double k, double chi,
+                                   double a, ccl_cl_tracer_collection_t *trc,
+                                   ccl_cosmology *cosmo,ccl_f2d_t *psp,
+                                   int *status) {
   int itr;
-  double transfer=0;
+  double transfer = 0;
 
-  for(itr=0;itr<trc->n_tracers;itr++) {
-    transfer+=transfer_limber_single(trc->ts[itr],l,lk,k,chi,a,cosmo,psp,status);
-    if(*status!=0)
+  for (itr=0; itr < trc->n_tracers; itr++) {
+    transfer += transfer_limber_single(
+      trc->ts[itr], l, lk, k, chi, a, cosmo, psp, status);
+    if (*status != 0)
       return -1;
   }
   return transfer;
 }
 
-static double cl_integrand(double lk,void *params)
-{
-  double d1,d2;
-  integ_cl_par *p=(integ_cl_par *)params;
-  double k=exp(lk);
-  double chi=(p->l+0.5)/k;
-  double a=ccl_scale_factor_of_chi(p->cosmo,chi,p->status);
-  d1=transfer_limber_wrap(p->l,lk,k,chi,a,p->trc1,
-			  p->cosmo,p->psp,p->status);
-  if(d1==0)
+static double cl_integrand(double lk, void *params) {
+  double d1, d2;
+  integ_cl_par *p = (integ_cl_par *)params;
+  double k = exp(lk);
+  double chi = (p->l+0.5)/k;
+  double a = ccl_scale_factor_of_chi(p->cosmo, chi, p->status);
+
+  d1 = transfer_limber_wrap(p->l, lk, k, chi, a, p->trc1,
+                            p->cosmo, p->psp, p->status);
+  if (d1 == 0)
     return 0;
-  d2=transfer_limber_wrap(p->l,lk,k,chi,a,p->trc2,
-			  p->cosmo,p->psp,p->status);
-  if(d2==0)
+
+  d2 = transfer_limber_wrap(p->l, lk, k, chi, a, p->trc2,
+                            p->cosmo, p->psp, p->status);
+
+  if (d2 == 0)
     return 0;
 
-  double pk=ccl_f2d_t_eval(p->psp,lk,a,p->cosmo,p->status);
+  double pk = ccl_f2d_t_eval(p->psp, lk, a, p->cosmo, p->status);
 
   return k*pk*d1*d2;
 }
 
-double ccl_angular_cl_limber(ccl_cosmology *cosmo,
-			     ccl_cl_tracer_collection_t *trc1,
-			     ccl_cl_tracer_collection_t *trc2,
-			     ccl_f2d_t *psp,double l,int *status)
-{
-  //Figure out which power spectrum to use
+void ccl_angular_cl_limber(ccl_cosmology *cosmo,
```

Comment:
should we rename this to `angular_cls_limber`?

Reply:
yup